clib/exninfo.ml

```diff
@@ -72,9 +72,15 @@ let get_backtrace e = get e backtrace_info
 
 let iraise (e,i) =
   let () = Mutex.lock lock in
-  let id = Thread.id (Thread.self ()) in
-  let () = current := (id, (e,i)) :: remove_assoc id !current in
-  let () = Mutex.unlock lock in
+  let () =
+  try
+    let id = Thread.id (Thread.self ()) in
+    let () = current := (id, (e,i)) :: remove_assoc id !current in
+    Mutex.unlock lock
+  with e -> (* Asynchronous exception *)
+    let () = Mutex.unlock lock in
+    raise e
```

Comment:
I wrote the same patch [but using a proper function `with_lock ~lock ~f` which IMHO should be preferred. However it seems to me that if the async exception happens after the update of the table we risk leaving `current` in an inconsistent state, isn't it the case?

Reply:
Thus it seems to me that handling `current` safely does require taking the global lock.

Reply:
`remove_assoc` and `find_and_remove_assoc` are stateless, so I do not see the problem: the modification to `current` is atomic. Or what inconsistent state are you speaking about?

Reply:
IIANM the modification to `current` will allocate, so timeouts can happen after and before the update, the cleanup code should take care of it.

Reply:
What is the problem with the timeout happening before or after the update?

Reply:
I tried to reason a bit about this, but however it turns out that already exninfo seems like completely broken in terms of semantics, should for example the metadata re-attached on re-raise? I will think a bit more but I don't think this patch is a regression on that area. For example we could get a timeout between the info and the usually later `iraise`, so current doesn't get cleaned up.

Regarding that particular code, I understand that in any case, sigalarm can only happen inside the lock before the imperative update is done, and not after, is that correct?

Reply:
Exninfo should only be used for debug, the claim that is completely broken misses the point. Half of the backtraces you added to the proof engine are semantically nonsense, but they are practically useful to get an idea of what went wrong. In any case, @gadmm is right that it doesn't matter, since asynchronous exceptions are outside of the backtrace model.

Reply:
> Half of the backtraces you added to the proof engine are semantically nonsense

You keep repeating this and even use "half", but TTBOMK your assertion doesn't hold, can you point out a single "nonsense" backtrace due to the recent changes there?

Reply:
And please stop being dismissive of other's work, specially when done as here, out of the blue, in a totally unrelated context, and actually with your argument being the exact opposite of the reality [before the exninfo work I did, _most_ of the backtraces were non-sense if in tactic code, not the opposite]

Reply:
The key reasoning here is:
- There is no data race with the signal handler
- Once an interrupt happens, the state of Exninfo for the current thread no longer matters (it can be any state as long as it is a valid state).

So as long as interrupts are outside your bracktrace thingy (`iraise` is not called from a signal handler, and the Exninfo is not queryed after an interrupt is caught), this looks correct to me.